vnext/Microsoft.ReactNative/CompositionViewComponent.idl

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "DocString.h"
+import "ViewProps.idl";
+import "IJSValueReader.idl";
+import "CompositionSwitcher.idl";
+
+namespace Microsoft.ReactNative.Composition
+{
+  [webhosthidden]
+  [experimental]
+  struct LayoutMetrics
+  {
+    Windows.Foundation.Rect Frame;
+    Single PointScaleFactor;
+  };
+
+  [webhosthidden]
+  [experimental] DOC_STRING(
+      "Interface to implement a custom view component based on View")
+  interface ICompositionViewComponent
```

Comment:
>ICompositionViewComponent [](http://example.com/codeflow?start=12&length=25)

This interface is too big and based on the TODOs inside is going to grow. The ideal interface to be implemented by custom code must have not more than one method. In many cases such interface can be replaced by a callback.

This is a foundation for the so called Builder pattern that we use for the Native Modules 2.0.

Everything that we want to say about the extension we provide to the builder interfaces. E.g., this is my initializer, this is how I can create visuals, this is how I can measure, etc.

Why the builder pattern is better? Extensibility. We can always add new methods to an interface that we implement, but we never can do it to the code implemented by third parties. As a result, for every single interface modification we must create a new interface and we start to see all these IViewComponent2, IViewComponentEx, IViewComponentAdv, IViewComponent3, etc.